README
NAME
    Device::CurrentCost - Perl modules for Current Cost energy monitors

VERSION
    version 1.142230

SYNOPSIS
      use Device::CurrentCost;
      my $envy = Device::CurrentCost->new(device => '/dev/ttyUSB0');

      $|=1; # don't buffer output

      while (1) {
        my $msg = $envy->read() or next;
        print $msg->summary, "\n";
      }

      use Device::CurrentCost::Constants;
      my $classic = Device::CurrentCost->new(device => '/dev/ttyUSB1',
                                             type => CURRENT_COST_CLASSIC);
      # ...

      open my $cclog, '<', 'currentcost.log' or die $!;
      my $cc = Device::CurrentCost->new(filehandle => $cclog);

      while (1) {
        my $msg = $cc->read() or next;
        print $msg->summary, "\n";
      }

DESCRIPTION
    Module for reading from Current Cost energy meters.

    IMPORTANT: This is an early release and the API is still subject to
    change.

    The API for history is definitely not complete. This will change soon
    and an mechanism for aggregating the history (which is split across many
    messages) should be added.

METHODS
  "new(%parameters)"
    This constructor returns a new Current Cost device object. The supported
    parameters are:

    device
        The name of the device to connect to. The value should be a tty
        device name, e.g. "/dev/ttyUSB0" but a pipe or plain file should
        also work. This parameter is mandatory if filehandle is not given.

    filehandle
        A filehandle to read from. This parameter is mandatory if device is
        not given.

    type
        The type of the device. Currently either "CURRENT_COST_CLASSIC" or
        "CURRENT_COST_ENVY". The default is "CURRENT_COST_ENVY".

    baud
        The baud rate for the device. The default is derived from the type
        and is either 57600 (for Envy) or 9600 (for classic).

    history_callback
        A function, taking a sensor id, a time interval and a hash reference
        of data as arguments, to be called every time a new complete set of
        history data becomes available. The data hash reference has keys of
        the number of intervals ago and values of the reading at that time.

  "device()"
    Returns the path to the device.

  "type()"
    Returns the type of the device.

  "baud()"
    Returns the baud rate.

  "filehandle()"
    Returns the filehandle being used to read from the device.

  "serial_port()"
    Returns the Device::SerialPort object for the device.

  "open()"
    This method opens the serial port and configures it.

  "read($timeout)"
    This method blocks until a new message has been received by the device.
    When a message is received a data structure is returned that represents
    the data received.

    IMPORTANT: This API is still subject to change.

  "read_one(\$buffer)"
    This method attempts to remove a single Current Cost message from the
    buffer passed in via the scalar reference. When a message is removed a
    data structure is returned that represents the data received. If
    insufficient data is available then undef is returned.

    IMPORTANT: This API is still subject to change.

  "sensor_history($sensor, $interval)"
    This method returns the most recent complete sensor data for the given
    sensor and the given interval (where interval must be one of 'hours',
    'days', 'months' or 'years'). The return value is a hash reference with
    keys for 'time' and 'data'. The 'time' value is the time (in seconds
    since epoch). The 'data' value is a hash reference with keys of the
    number of intervals ago and values of the reading at that time.

    It returns undef if no history data has been received yet.

AUTHOR
    Mark Hindess <soft-cpan@temporalanomaly.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Mark Hindess.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Device::CurrentCost.

{{$NEXT}}

1.232151  2023-08-03 21:02:46+01:00 Europe/London

1.232150  2023-08-03 20:58:24+01:00 Europe/London
    - Fix Device::SerialPort imports (thx @haarg).

1.142240  2014-08-12 15:46:29+01:00 Europe/London
    - Fix constant usage in tests to avoid error on perl 5.6.x.

1.142230  2014-08-11 20:49:13+01:00 Europe/London
    - Fix constant usage to avoid error on perl 5.6.x.
    - Fix test warning.

1.130190  2013-01-19 17:27:09 Europe/London
    - Use Device::SerialPort.

1.112970  2011-10-24 22:16:21 Europe/London
    - Modify timeout test to make it less error prone.
    - Pod link fix.

1.110792  2011-03-20 21:17:26 Europe/London
    - Make reader parser more liberal so it accepts more types of
      input.

1.110791  2011-03-20 13:03:26 Europe/London
    - Minor regexp improvement.

1.110790  2011-03-20 09:19:26 Europe/London
    - Add history API.

1.110390  2011-02-08 22:59:34 Europe/London
    - Fix tests on non-linux platforms.

1.110370  2011-02-06 20:35:11 Europe/London
    - Initial release.


t/Helpers.pm
package t::Helpers;

=head1 NAME

t::Helpers - Perl extension for Helper functions for tests.

=head1 SYNOPSIS

  use Test::More tests => 2;
  use t::Helpers qw/:all/;
  is(test_error(sub { die 'argh' }),
     'argh',
     'died horribly');

  is(test_warn(sub { warn 'danger will robinson' }),
     'danger will robinson',
     'warned nicely');

=head1 DESCRIPTION

Common functions to make test scripts a bit easier to read.  There are
CPAN modules to do this sort of thing, but most people wont have them
installed and they are pretty trivial functions so to encourage
testing they are included here.

=cut

use 5.006;
use strict;
use warnings;

use Exporter;

our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [ qw(
                                    test_error
                                    test_warn
) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();
our $VERSION = 0.1;

=head2 C<test_error($code_ref)>

This method runs the code with eval and returns the error.  It strips
off some common strings from the end of the message including any "at
<file> line <number>" strings and any "(@INC contains: .*)".

=cut

sub test_error {
  my $sub = shift;
  eval { $sub->() };
  my $error = $@;
  if ($error) {
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $error;
}

=head2 C<test_warn($code_ref)>

This method runs the code with eval and returns the warning.  It strips
off any "at <file> line <number>" specific part(s) from the end.

=cut

sub test_warn {
  my $sub = shift;
  my $warn;
  local $SIG{__WARN__} = sub { $warn .= $_[0]; };
  eval { $sub->(); };
  die $@ if ($@);
  if ($warn) {
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $warn;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.030.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl modules for Current Cost energy monitors",
  "AUTHOR" => "Mark Hindess <soft-cpan\@temporalanomaly.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Device-CurrentCost",
  "EXE_FILES" => [
    "bin/current-cost-reader"
  ],
  "LICENSE" => "perl",
  "NAME" => "Device::CurrentCost",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Device::SerialPort" => 0,
    "Fcntl" => 0,
    "Getopt::Long" => 0,
    "IO::Handle" => 0,
    "IO::Select" => 0,
    "List::Util" => 0,
    "Pod::Usage" => 0,
    "Symbol" => 0,
    "Time::HiRes" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Pipe" => 0,
    "Test::More" => 0,
    "Test::Requires" => 0,
    "Test::SharedFork" => 0,
    "Test::Warn" => 0,
    "lib" => 0
  },
  "VERSION" => "1.232150",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Device::SerialPort" => 0,
  "ExtUtils::MakeMaker" => 0,
  "Fcntl" => 0,
  "File::Spec" => 0,
  "Getopt::Long" => 0,
  "IO::Handle" => 0,
  "IO::Pipe" => 0,
  "IO::Select" => 0,
  "List::Util" => 0,
  "Pod::Usage" => 0,
  "Symbol" => 0,
  "Test::More" => 0,
  "Test::Requires" => 0,
  "Test::SharedFork" => 0,
  "Test::Warn" => 0,
  "Time::HiRes" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Device/CurrentCost/Message.pm
use strict;
use warnings;
package Device::CurrentCost::Message;
$Device::CurrentCost::Message::VERSION = '1.232150';
# ABSTRACT: Perl modules for Current Cost energy monitor messages


use constant DEBUG => $ENV{DEVICE_CURRENT_COST_DEBUG};

use Carp qw/croak carp/;
use Device::CurrentCost::Constants;
use List::Util qw/min/;


sub new {
  my ($pkg, %p) = @_;
  croak $pkg.'->new: message parameter is required' unless (exists $p{message});
  my $self = bless { %p }, $pkg;
  $self;
}


sub device_type {
  my $self = shift;
  return $self->{device_type} if (exists $self->{device_type});
  $self->{device_type} =
    $self->message =~ m!<src><name>! ? CURRENT_COST_CLASSIC : CURRENT_COST_ENVY;
}


sub device {
  my $self = shift;
  return $self->{device}->[0] if (exists $self->{device});
  $self->_find_device->[0]
}


sub device_version {
  my $self = shift;
  return $self->{device}->[1] if (exists $self->{device});
  $self->_find_device->[1]
}

sub _find_device {
  my $self = shift;
  my $name = $self->_parse_field('name');
  $self->{device} =
    defined $name ?
      [ $name, 'v'.$self->_parse_field('sver')] :
        [ split /-/, $self->_parse_field('src'), 2 ];
}


sub message { shift->{message} }

sub _parse_field {
  my ($self, $field, $default) = @_;
  return $self->{$field} if (exists $self->{$field});
  if ($self->message =~ m!<$field>(.*?)</$field>!s) {
    my $v = $1;
    $self->{$field} =
      ($v =~ m!<([^>]+)>(.*?)</\1>!s) ? { value => $2, units => $1 } : $v;
  } elsif (defined $default) {
    return $default;
  } else {
    return
  }
}


sub dsb { shift->_parse_field('dsb') }



sub days_since_boot { shift->dsb }


sub time {
  my $self = shift;
  my $time = $self->_parse_field('time');
  return $time if (defined $time);
  $self->{time} =
    $self->_parse_field('hr').':'.
      $self->_parse_field('min').':'.
        $self->_parse_field('sec')
}


sub time_in_seconds {
  my $self = shift;
  return $self->{time_in_seconds} if (exists $self->{time_in_seconds});
  my ($h, $m, $s) = split /:/, $self->time, 3;
  $self->{time_in_seconds} = $h*3600 + $m*60 + $s;
}


sub boot_time {
  my $self = shift;
  $self->days_since_boot * 86400 + $self->time_in_seconds
}


sub sensor { shift->_parse_field('sensor', 0) }


sub id { shift->_parse_field('id') }


sub type { shift->_parse_field('type') }


sub tmpr { shift->_parse_field('tmpr') }


sub temperature { shift->tmpr }


sub has_history { shift->message =~ /<hist>/ }


sub has_readings { shift->message =~ /<ch1>/ }


sub units {
  my $self = shift;
  my $ch1 = $self->_parse_field('ch1') or return;
  $ch1->{units}
}


sub value {
  my ($self, $channel) = @_;
  $self->units || return; # return if no units can be found - historic only
  if ($channel) {
    return $self->_parse_field('ch'.$channel, { value => undef })->{value};
  }
  return $self->{total} if (exists $self->{total});
  foreach (1 .. 3) {
    $self->{total} += $self->value($_)||0;
  }
  $self->{total}
}


sub summary {
  my ($self, $prefix) = @_;
  $prefix = '' unless (defined $prefix);
  my $str = $prefix.'Device: '.$self->device.' '.$self->device_version."\n";
  $prefix .= '  ';
  if ($self->has_readings) {
    $str .= $prefix.'Sensor: '.$self->sensor;
    $str .= (' ['.$self->id.','.$self->type."]\n".
             $prefix.'Total: '.$self->value.' '.$self->units."\n");
    foreach my $phase (1..3) {
      my $v = $self->value($phase);
      next unless (defined $v);
      $str .= $prefix.'Phase '.$phase.': '.($v+0)." ".$self->units."\n";
    }
  }
  if ($self->has_history) {
    $str .= $prefix."History\n";
    my $hist = $self->history;
    foreach my $sensor (sort keys %$hist) {
      $str .= $prefix.'  Sensor '.$sensor."\n";
      foreach my $span (sort keys %{$hist->{$sensor}}) {
        foreach my $age (sort { $a <=> $b } keys %{$hist->{$sensor}->{$span}}) {
          $str .= $prefix.'    -'.$age.' '.$span.': '.
            (0+$hist->{$sensor}->{$span}->{$age})."\n";
        }
      }
    }
  }
  $str
}


sub history {
  my $self = shift;
  return $self->{history} if (exists $self->{history});
  my %hist = ();
  $self->{history} = \%hist;
  return $self->{history} unless ($self->has_history);
  my $xml = $self->message;
  if ($xml =~ /<data>/) {
    # envy
    foreach my $data (split qr!</data><data>!, $xml) {
      my ($sensor) = ($data =~ /<sensor>(\d+)</) or next;
      my %rec = ();
      $hist{$sensor} = _parse_history($data);
    }
  } else {
    # classic
    $hist{$self->sensor} = _parse_history($xml);
  }
  \%hist;
}

sub _parse_history {
  my $string = shift;
  my %rec = ();
  foreach my $span (qw/hours days months years/) {
    my $first = substr $span, 0, 1;
    while ($string =~ m!<$first(\d+)>([^<]+)</$first\1>!mg) {
      $rec{$span}->{0+$1} = 0+$2;
    }
  }
  \%rec;
}

1;

=pod

=encoding UTF-8

=head1 NAME

Device::CurrentCost::Message - Perl modules for Current Cost energy monitor messages

=head1 VERSION

version 1.232150

=head1 SYNOPSIS

  use Device::CurrentCost::Message;
  my $msg = Device::CurrentCost::Message->new(message => '<msg>...</msg>');
  print 'Device: ', $msg->device, ' ', $msg->device_version, "\n";
  if ($msg->has_readings) {
    print 'Sensor: ', $msg->sensor, '.', $msg->id, ' (', $msg->type, ")\n";
    print 'Total: ', $msg->value, ' ', $msg->units, "\n";
    foreach my $phase (1..3) {
      print 'Phase ', $phase, ': ',
        $msg->value($phase)+0, " ", $msg->units, "\n";
    }
  }

  use Data::Dumper;
  print Data::Dumper->Dump([$msg->history]) if ($msg->has_history);

  # or
  print $msg->summary, "\n";

=head1 DESCRIPTION

=head1 METHODS

=head2 C<new(%parameters)>

This constructor returns a new Current Cost message object.
The supported parameters are:

=over

=item message

The message data.  Usually a string like 'C<< <msg>...</msg> >>'.
This parameter is required.

=back

=head2 C<device_type()>

Returns the type of the device that created the message.

=head2 C<device()>

Returns the name of the device that created the message.

=head2 C<device_version()>

Returns the version of the device that created the message.

=head2 C<message()>

Returns the raw data of the message.

=head2 C<dsb()>

Returns the days since boot field of the message.

=head2 C<days_since_boot()>

Returns the days since boot field of the message.

=head2 C<time()>

Returns the time field of the message in C<HH:MM:SS> format.

=head2 C<time_in_seconds()>

Returns the time field of the message in seconds.

=head2 C<boot_time()>

Returns the time since boot reported by the message in seconds.

=head2 C<sensor()>

Returns the sensor number field of the message.  A classic monitor
supports only one sensor so 0 is returned.

=head2 C<id()>

Returns the id field of the message.

=head2 C<type()>

Returns the sensor type field of the message.

=head2 C<tmpr()>

Returns the tmpr/temperature field of the message.

=head2 C<temperature()>

Returns the temperature field of the message.

=head2 C<has_history()>

Returns true if the message contains history data.

=head2 C<has_readings()>

Returns true if the message contains current data.

=head2 C<units()>

Returns the units of the current data readings in the message.

=head2 C<value( [$channel] )>

Returns the value of the current data reading for the given channel
(phase) in the message.  If no channel is given then the total of all
the current data readings for all channels is returned.

=head2 C<summary( [$prefix] )>

Returns the string summary of the data in the message.  Each line of the
string is prefixed by the given prefix or the empty string if the prefix
is not supplied.

=head2 C<history()>

Returns a data structure contain any history data from the message.

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

__END__

ENVY:

<msg><src>CC128-v0.11</src><dsb>00089</dsb><time>13:02:39</time><tmpr>18.7</tmpr><sensor>1</sensor><id>01234</id><type>1</type><ch1><watts>00345</watts></ch1><ch2><watts>02151</watts></ch2><ch3><watts>00000</watts></ch3></msg>

<msg><src>CC128-v0.11</src><dsb>00596</dsb><time>13:11:20</time><hist><dsw>00597</dsw><type>1</type><units>kwhr</units><data><sensor>0</sensor><h250>7.608</h250><h248>7.163</h248><h246>6.541</h246><h244>3.270</h244></data><data><sensor>1</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>2</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>3</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>4</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>5</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>6</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>7</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>8</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data><data><sensor>9</sensor><h250>0.000</h250><h248>0.000</h248><h246>0.000</h246><h244>0.000</h244></data></hist></msg>

CLASSIC:

<msg><date><dsb>00001</dsb><hr>12</hr><min>32</min><sec>01</sec></date><src><name>CC02</name><id>12345</id><type>1</type><sver>1.06</sver></src><ch1><watts>07806</watts></ch1><ch2><watts>00144</watts></ch2><ch3><watts>00144</watts></ch3><tmpr>21.1</tmpr></msg>

<msg><date><dsb>00001</dsb><hr>12</hr><min>32</min><sec>13</sec></date><src><name>CC02</name><id>12345</id><type>1</type><sver>1.06</sver></src><ch1><watts>07752</watts></ch1><ch2><watts>00144</watts></ch2><ch3><watts>00144</watts></ch3><tmpr>21.0</tmpr><hist><hrs><h02>001.3</h02><h04>000.0</h04><h06>000.0</h06><h08>000.0</h08><h10>000.0</h10><h12>000.0</h12><h14>000.0</h14><h16>000.0</h16><h18>000.0</h18><h20>000.0</h20><h22>000.0</h22><h24>000.0</h24><h26>000.0</h26></hrs><days><d01>0000</d01><d02>0000</d02><d03>0000</d03><d04>0000</d04><d05>0000</d05><d06>0000</d06><d07>0000</d07><d08>0000</d08><d09>0000</d09><d10>0000</d10><d11>0000</d11><d12>0000</d12><d13>0000</d13><d14>0000</d14><d15>0000</d15><d16>0000</d16><d17>0000</d17><d18>0000</d18><d19>0000</d19><d20>0000</d20><d21>0000</d21><d22>0000</d22><d23>0000</d23><d24>0000</d24><d25>0000</d25><d26>0000</d26><d27>0000</d27><d28>0000</d28><d29>0000</d29><d30>0000</d30><d31>0000</d31></days><mths><m01>0000</m01><m02>0000</m02><m03>0000</m03><m04>0000</m04><m05>0000</m05><m06>0000</m06><m07>0000</m07><m08>0000</m08><m09>0000</m09><m10>0000</m10><m11>0000</m11><m12>0000</m12></mths><yrs><y1>0000000</y1><y2>0000000</y2><y3>0000000</y3><y4>0000000</y4></yrs></hist></msg>


lib/Device/CurrentCost/Constants.pm
use strict;
use warnings;
package Device::CurrentCost::Constants;
$Device::CurrentCost::Constants::VERSION = '1.232150';
# ABSTRACT: Module to export constants for Current Cost devices


my %constants =
  (
   CURRENT_COST_CLASSIC => 0x1,
   CURRENT_COST_ENVY => 0x2,
  );
my %names =
  (
   $constants{CURRENT_COST_ENVY} => 'Envy',
   $constants{CURRENT_COST_CLASSIC} => 'Classic',
  );

sub import {
  no strict qw/refs/; ## no critic
  my $pkg = caller(0);
  foreach (keys %constants) {
    my $v = $constants{$_};
    *{$pkg.'::'.$_} = sub () { $v };
  }
  foreach (qw/current_cost_type_string/) {
    *{$pkg.'::'.$_} = \&{$_};
  }
}


sub current_cost_type_string {
  $names{$_[0]}
}

__END__

=pod

=encoding UTF-8

=head1 NAME

Device::CurrentCost::Constants - Module to export constants for Current Cost devices

=head1 VERSION

version 1.232150

=head1 SYNOPSIS

  use Device::CurrentCost::Constants;

=head1 DESCRIPTION

Module to export constants for Current Cost devices

=head1 C<FUNCTIONS>

=head2 C<current_cost_type_string( $type )>

Returns a string describing the given Current Cost device type.

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Device/CurrentCost.pm
use strict;
use warnings;
package Device::CurrentCost;
$Device::CurrentCost::VERSION = '1.232150';
# ABSTRACT: Perl modules for Current Cost energy monitors


use constant DEBUG => $ENV{DEVICE_CURRENT_COST_DEBUG};

use Carp qw/croak carp/;
use Device::CurrentCost::Constants;
use Device::CurrentCost::Message;
use Device::SerialPort;
use Fcntl;
use IO::Handle;
use IO::Select;
use Symbol qw(gensym);
use Time::HiRes;


sub new {
  my ($pkg, %p) = @_;
  my $self = bless {
                    buf => '',
                    discard_timeout => 1,
                    type => CURRENT_COST_ENVY,
                    history_callback => sub {},
                    %p
                   }, $pkg;
  croak $pkg.q{->new: 'device' parameter is required}
    unless (exists $p{device} or exists $p{filehandle});
  $self->open();
  $self;
}


sub device { shift->{device} }


sub type { shift->{type} }


sub baud {
  my $self = shift;
  defined $self->{baud} ? $self->{baud} :
    $self->type == CURRENT_COST_CLASSIC ? 9600 : 57600;
}


sub filehandle { shift->{filehandle} }


sub serial_port { shift->{serial_port} }


sub open {
  my $self = shift;

  my $fh = $self->filehandle;
  unless ($fh) {
    my $dev = $self->device;
    print STDERR 'Opening serial port: ', $dev, "\n" if DEBUG;
    my $fh = gensym();
    my $s = tie *$fh, 'Device::SerialPort', $dev or
      croak "Could not tie serial port, $dev, to file handle: $!";
    foreach my $setting ([ baudrate => $self->baud ],
                         [ databits => 8 ],
                         [ parity => 'none' ],
                         [ stopbits => 1 ],
                         [ datatype => 'raw' ]) {
      my ($setter, @v) = @$setting;
      $s->$setter(@v);
    }
    $s->write_settings();
    sysopen($fh, $dev, O_RDWR|O_NOCTTY|O_NDELAY) or
      croak "sysopen of '$dev' failed: $!";
    $self->{serial_port} = $s;
    $self->{filehandle} = $fh;
  }
  $self->filehandle;
}


sub read {
  my ($self, $timeout) = @_;
  my $res = $self->read_one(\$self->{buf});
  return $res if (defined $res);
  $self->_discard_buffer_check();
  my $fh = $self->filehandle;
  my $sel = IO::Select->new($fh);
  do {
    my $start = $self->_time_now;
    $sel->can_read($timeout) or return;
    my $bytes = sysread $fh, $self->{buf}, 2048, length $self->{buf};
    $self->{_last_read} = $self->_time_now;
    $timeout -= $self->{_last_read} - $start if (defined $timeout);
    unless ($bytes) {
      croak((ref $self).'->read: '.(defined $bytes ? 'closed' : 'error: '.$!));
    }
    print STDERR 'Read ', $bytes, "bytes\n" if DEBUG;
    $res = $self->read_one(\$self->{buf});
    return $res if (defined $res);
  } while (1);
}


sub read_one {
  my ($self, $rbuf) = @_;
  return unless ($$rbuf);
  print STDERR 'Read one from !', $$rbuf, "!\n" if DEBUG;
  if ($$rbuf =~ s!^.*?(<msg>.*?</msg>)[\r\n ]*!!s) {
    my $msg = Device::CurrentCost::Message->new(message => $1);
    my $t = $self->_time_now;
    if ($msg->has_history) {
      my $new = $msg->history;
      my $our= $self->{history} || ($self->{history} = {});
      foreach my $sensor (sort keys %$new) {
        foreach my $interval (sort keys %{$new->{$sensor}}) {
          foreach my $age (keys %{$new->{$sensor}->{$interval}}) {
            $our->{$sensor}->{$interval}->{pending}->{$age} =
              0+$new->{$sensor}->{$interval}->{$age};
          }
          if (exists $our->{$sensor}->{$interval}->{pending}->{1} ||
              ($interval eq 'hours' &&
               (exists $our->{$sensor}->{$interval}->{pending}->{4} ||
                exists $our->{$sensor}->{$interval}->{pending}->{2}))) {
            my $entries = keys %{$our->{$sensor}->{$interval}->{pending}};
            if ($entries == { years => 4, months => 21, # envy
                              days => 90, hours => 372 }->{$interval} ||
                $entries == { years => 4, months => 12, # classic
                              days => 31, hours => 13 }->{$interval}) {
              %{$our->{$sensor}->{$interval}->{current}} =
                %{$our->{$sensor}->{$interval}->{pending}};
              $our->{$sensor}->{$interval}->{time} = $t;
              $self->{history_callback}->($sensor, $interval,
                   $our->{$sensor}->{$interval}->{current});
            }
            $our->{$sensor}->{$interval}->{pending} = {};
          }
        }
      }
    }
    return $msg;
  } else {
    return;
  }
}


sub sensor_history {
  my ($self, $sensor, $interval) = @_;
  return unless (exists $self->{history}->{$sensor}->{$interval}->{current});
  return {
          time => $self->{history}->{$sensor}->{$interval}->{time},
          data => $self->{history}->{$sensor}->{$interval}->{current}
         };
}

sub _discard_buffer_check {
  my $self = shift;
  if ($self->{buf} ne '' &&
      $self->{_last_read} < ($self->_time_now - $self->{discard_timeout})) {
    carp "Discarding '", $self->{buf}, "'";
    $self->{buf} = '';
  }
}

sub _time_now {
  Time::HiRes::time;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Device::CurrentCost - Perl modules for Current Cost energy monitors

=head1 VERSION

version 1.232150

=head1 SYNOPSIS

  use Device::CurrentCost;
  my $envy = Device::CurrentCost->new(device => '/dev/ttyUSB0');

  $|=1; # don't buffer output

  while (1) {
    my $msg = $envy->read() or next;
    print $msg->summary, "\n";
  }

  use Device::CurrentCost::Constants;
  my $classic = Device::CurrentCost->new(device => '/dev/ttyUSB1',
                                         type => CURRENT_COST_CLASSIC);
  # ...

  open my $cclog, '<', 'currentcost.log' or die $!;
  my $cc = Device::CurrentCost->new(filehandle => $cclog);

  while (1) {
    my $msg = $cc->read() or next;
    print $msg->summary, "\n";
  }

=head1 DESCRIPTION

Module for reading from Current Cost energy meters.

B<IMPORTANT:> This is an early release and the API is still subject to
change.

The API for history is definitely not complete.  This will change soon
and an mechanism for aggregating the history (which is split across
many messages) should be added.

=head1 METHODS

=head2 C<new(%parameters)>

This constructor returns a new Current Cost device object.  The
supported parameters are:

=over

=item device

The name of the device to connect to.  The value should be a tty
device name, e.g. C</dev/ttyUSB0> but a pipe or plain file should also
work.  This parameter is mandatory if B<filehandle> is not given.

=item filehandle

A filehandle to read from.  This parameter is mandatory if B<device> is
not given.

=item type

The type of the device.  Currently either C<CURRENT_COST_CLASSIC> or
C<CURRENT_COST_ENVY>.  The default is C<CURRENT_COST_ENVY>.

=item baud

The baud rate for the device.  The default is derived from the type and
is either C<57600> (for Envy) or C<9600> (for classic).

=item history_callback

A function, taking a sensor id, a time interval and a hash reference
of data as arguments, to be called every time a new complete set of
history data becomes available.  The data hash reference has keys of
the number of intervals ago and values of the reading at that time.

=back

=head2 C<device()>

Returns the path to the device.

=head2 C<type()>

Returns the type of the device.

=head2 C<baud()>

Returns the baud rate.

=head2 C<filehandle()>

Returns the filehandle being used to read from the device.

=head2 C<serial_port()>

Returns the Device::SerialPort object for the device.

=head2 C<open()>

This method opens the serial port and configures it.

=head2 C<read($timeout)>

This method blocks until a new message has been received by the
device.  When a message is received a data structure is returned
that represents the data received.

B<IMPORTANT:> This API is still subject to change.

=head2 C<read_one(\$buffer)>

This method attempts to remove a single Current Cost message from the
buffer passed in via the scalar reference.  When a message is removed
a data structure is returned that represents the data received.  If
insufficient data is available then undef is returned.

B<IMPORTANT:> This API is still subject to change.

=head2 C<sensor_history($sensor, $interval)>

This method returns the most recent complete sensor data for the
given sensor and the given interval (where interval must be one
of 'hours', 'days', 'months' or 'years').  The return value is
a hash reference with keys for 'time' and 'data'.  The 'time'
value is the time (in seconds since epoch).  The 'data' value
is a hash reference with keys of the number of intervals ago
and values of the reading at that time.

It returns undef if no history data has been received yet.

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
